src/lib/FreeC/Pass/SharingAnalysisPass.hs

```diff
@@ -0,0 +1,144 @@
+--
+--   Computes @let@-bindings from the given variables, composes the resulting
+--   substitutions and applies the substitution on the expression.
+buildLet
+    :: IR.Expr -> [ ( IR.VarName, Maybe IR.TypeScheme ) ] -> Converter IR.Expr
+buildLet expr []   = return expr
+buildLet expr vars = do
+    let srcSpan = IR.exprSrcSpan expr
+    ( binds, substs ) <- buildBinds srcSpan vars
+    return (IR.Let srcSpan binds (applySubst (composeSubsts substs) expr)
+            (IR.exprTypeScheme expr))
+
+-- | Converts the list containing variables into @let@-bindings where
+--   the variable pattern is a fresh variable and the right-hand side is
+--   a variable that occurred multiple times on right hand sides.
+--
+--   Also computes substitutions mapping given variables to fresh variables.
+--   Returns the generated let-bindings and the substitutions.
+buildBinds :: SrcSpan -> [ ( IR.VarName, Maybe IR.TypeScheme ) ]
+    -> Converter ( [ IR.Bind ], [ Subst IR.Expr ] )
+buildBinds srcSpan = foldM buildBind ( [], [] )
+  where
+    buildBind :: ( [ IR.Bind ], [ Subst IR.Expr ] )
+        -> ( IR.VarName, Maybe IR.TypeScheme )
+        -> Converter ( [ IR.Bind ], [ Subst IR.Expr ] )
+    buildBind ( binds, substs ) ( varName, varTypeScheme ) = do
+        varName' <- freshHaskellQName varName
+        let subst = singleSubst' varName (\s -> IR.Var s varName' varTypeScheme)
+            -- This type scheme has wrong srcSpanInformation
+            rhs = IR.Var srcSpan varName varTypeScheme
+            Just varPatIdent = IR.identFromQName varName'
+            varPatType = fmap IR.typeSchemeType varTypeScheme
+            varPat = IR.VarPat srcSpan varPatIdent varPatType False
+            bind = IR.Bind srcSpan varPat rhs
+        _ <- renameAndDefineVar srcSpan False varPatIdent varPatType
+        return ( bind : binds, subst : substs )
+
+-- | Counts all variable names on right-hand sides of expression.
+--
+--   If a variable is annotated with a type scheme the annotation is introduced
+--   into the map as well.
+countVarNames :: IR.Expr -> Map IR.VarName ( Integer, Maybe IR.TypeScheme )
+countVarNames IR.Con {} = Map.empty
+countVarNames (IR.Var _ varName maybeTypeScheme)
+    = Map.singleton varName ( 1, maybeTypeScheme )
+countVarNames (IR.App _ lhs rhs _)
+    = countVarNames lhs `mergeMap` countVarNames rhs
+countVarNames (IR.TypeAppExpr _ lhs _ _) = countVarNames lhs
+countVarNames (IR.If _ e1 e2 e3 _) = (countVarNames e1)
+    `mergeMap` (countVarNames e2) `mergeMap` countVarNames e3
+countVarNames (IR.Case _ e alts _) = countVarNames e `mergeMap` foldr mergeMap
+    Map.empty (map (countVarNames . IR.altRhs) alts)
+countVarNames IR.Undefined {} = Map.empty
+countVarNames IR.ErrorExpr {} = Map.empty
+countVarNames IR.IntLiteral {} = Map.empty
+countVarNames (IR.Lambda _ _ rhs _) = countVarNames rhs
+countVarNames (IR.Let _ binds e _) = countVarNames e `mergeMap` foldr mergeMap
+    Map.empty (map (countVarNames . IR.bindExpr) binds)
```

Comment:
`case`-expressions alternatives, lambda expressions and `let`-bindings can shadow variables. Shadowed variables should probably not be counted.

Reply:
Good catch, i didn't think of that.